### ISSUE TITLE
Error viewing article in portfolio theme

### ISSUE THREAD

**Issue by wu-lee** [pencilblue/pencilblue#1182]
**Description:**

500 error on a new site when viewing a newly created article when the portfolio theme is active.

**Steps to Reproduce:**

(Note, This is as I have done so far, I've not had time to verify this is repeatable, and there have been diversions)

- install pencilblue (multi-tenant mode)
- enable portfolio plugin
- enable portfolio theme
- create a new article called "An Article" with the default generated url (/article/An-article)
- save it
- view it (need to type in the url, as there is no navigation link)

The result for me is a 500 error page, with the following stack trace

    TypeError: Cannot read property 'article_topics' of undefined
	at /home/pencilblue/src/plugins/portfolio/controllers/blog.js:442:37
	at /home/pencilblue/src/include/service/entities/plugin_setting_service.js:108:13
	at /home/pencilblue/src/include/service/simple_layered_service.js:95:17
	at /home/pencilblue/src/node_modules/async/lib/async.js:789:30
	at /home/pencilblue/src/node_modules/async/lib/async.js:167:37
	at /home/pencilblue/src/include/service/simple_layered_service.js:76:21
	at callback (/home/pencilblue/src/include/service/db_entity_service.js:92:13)
	at /home/pencilblue/src/include/dao/dao.js:144:12
	at /home/pencilblue/src/include/dao/dao.js:272:17
	at handleCallback (/home/pencilblue/src/node_modules/mongodb/lib/utils.js:96:12)

**Configuration Details:**

    module.exports = {
	siteName: "<redacted name>",
	siteRoot: "http://<redacted url>",
	sitePort: 8700,
	siteIP: "127.0.0.1",
	logging: {
	    level: "silly"
	},
	db: {
	    type:"mongo",
	    servers: [
	      "127.0.0.1:27017"
	    ],
	    name: "pencilblue",
	    writeConcern: 1
	},
	cache: {
	    fake: true,
	    host: "localhost",
	    port: 6379
	},
	settings: {
	    use_memory: false,
	    use_cache: false
	},
	templates: {
	    use_memory: true,
	    use_cache: false
	},
	plugins: {
	    caching: {
		use_memory: false,
		use_cache: false
	    }
	},
	command: {
	    broker: "mongo"
	},
	registry: {
	    type: "mongo"
	},
	session: {
	    storage: "mongo"
	},
	media: {
	    provider: "mongo",
	    max_upload_size: 6 * 1024 * 1024
	},
	cluster: {
	    workers: 1,
	    self_managed: true
	},
	multisite: {
	    enabled: true,
	    globalRoot: 'http://localhost:8700'
	}
    };

**Log Output:**

(Look for "An-article" and "An error occurred during controller execution")
[snippet.log.txt](https://github.com/pencilblue/pencilblue/files/626835/snippet.log.txt)

PencilBlue Version: 0.7.0, master branch commit 370058e0dc83

Node.js Version: v4.6.2

Environment Description:

Firefox 50.0 on Ubuntu precise.
Debian "Wheezy" webserver.

**Comment by wu-lee** [comment]
Note, if I switch back to the pencilblue theme, this error goes away when I view the article.  Relevant log portion attached.

[snippet.log.txt](https://github.com/pencilblue/pencilblue/files/626863/snippet.log.txt)

**Comment by brianhyder** [comment]
@wu-lee I went through your steps and I could not reproduce.  I have a couple of questions about your setup.

1. When you went through your setup process did you have multi-tenancy enabled?
2. In your configuration, was your siteRoot the same as your global root?
3. In your steps to reproduce it didn't mention that you activated a site.  Did you go through that process?
4. Did you enable the portfolio theme on the global site or the child site?

**Comment by wu-lee** [comment]
1.  Yes, although I didn't use pbctrl, I installed from git, npm installed, and edited the config by hand before running `node pencilblue.js` (usually via an init script).
2. No, siteRoot was set to a publicly accessible domain, reverse-proxied to the nodejs server.  globalRoot was set to http://localhost:8700,  which I accessed via an ssh tunnel (when necessary, which seemed only for the initial visit).
3. Probably on the initial visit, although I confess I'd forgotten that step.  If I recall correctly, there was a single site as defined by siteRoot/siteName etc. pre-existing.  From then on the admin panel seemed to be accessible on the child sites, which I created a second of.
4. The portfolio theme was enabled on the initial child site defined by siteName.

**Comment by wu-lee** [comment]
Ok, managed to replicate.  Too several goes to get it off perfectly without mistakes.

I'm installing with ansible, but the procedure and config is exactly as above, except without:

    command: {
        broker: "mongo"
    },

The version installed is current master, i.e. commit c42de2421270.

I can confirm the public sites are not visible until I run the set up.  To do this I visit the global URL via an ssh tunnel, and create my admin account.

I log in, then enable the site defined in the config.  (The first attempt the progress bar halts at 50% again, although the site was enabled after a page refresh. This works correctly on my last attempt. Not certain why, but the first attempt had config mistakes with the `siteRoot` being http://example.com and and `mutisite.enabled` being false).

I log into the child site as admin via the /admin path.

I enable the portfolio plugin (dialog behaves this time, closes self)

I enable portfolio theme (dialog doesn't seem to self close, so I click OK.)

I create a new article as above, title 'An-account', URL `/article/An-article`, content "An article content".  Saved (not as save draft).

I visited the URL.  And I can confirm the same error, exactly as above.

However, trying again later, after writing this (a few minutes), the article gets rendered with no error.

The `silly` level log is attached here, which seems to show lots of stuff going on after my initial page visit.  Note, I don't visit the new page for about 20 seconds.  However, if time resolutions are 1 minute, perhaps this is too quick!

[snippet.log.txt](https://github.com/pencilblue/pencilblue/files/635151/snippet.log.txt)

**Comment by brianhyder** [comment]
I'm honestly baffled why this isn't working for you.  The weird part is that you are saying that if you wait it works.  This leads me to believe that your MongoDB instance is not writing the record fast enough.  Are you using a remote mongo service or is it a local instance to your installation of PB?

I would like to try one last thing to see if we can get this working:
In your config.js file set ```db.query_logging=true```.  This will print out every query that is being executed against Mongo.  The error is being caused by the fact that there is no content being returned by the query.  Making sure the query is what we expect would be the next step.  

Are you only able to create the error in your remote environment or can you reproduce it locally too?

**Comment by wu-lee** [comment]
My MongoDB instance is local - it's all on the same server, which happens to be a VM running at an ISP.  I know VMs sometimes have timing issues, not sure if that's going to be relevant, and I've never noticed any problems with it so far.

I'll give the query logging a try later.  So far I've not really used pencil blue except on the VM, apart from a quick sanity check early on.  I'll also try replicating the problem elsewhere tho.

**Comment by brianhyder** [comment]
Thank you. I know tracking this down hasn't been quick or easy. The help is
greatly appreciated. It gives us insight into the areas where code or docs
are lacking.

**Comment by wu-lee** [comment]
Tried again on my laptop - Runs Ubuntu Precise, with nodejs 4.7.0, mongodb 1:2.0.4-1ubuntu2.1.

This appeared to work fine, using exactly the same set-up and process as above, although manual and with siteName and siteRoot modified to work with my laptop's hostname.  Global root unchanged.

(Aside: version of mongo on Wheezy VM is 1:2.0.6-1.1+deb7u1)

Log attached.
[snippet.log.txt](https://github.com/pencilblue/pencilblue/files/637892/snippet.log.txt)

**Comment by brianhyder** [comment]
Ok... that tells me it is less likely to be a code error and more likely to be environmental.  My suggestion at this point is to turn on query logging and make sure that the query that retrieves the article to be rendered has the correct ID and site.  Additionally, you could add ```console.log``` statements directly into the portfolio theme but at this time I'm not sure there is much more help we can provide.

**Comment by wu-lee** [comment]
In another inconclusive attempt to work out what's going on,  I notice that my server seems to be respawining workers, as per the attached log section.  This was after inserting trace statements, restarting pencilblue, then creating a new article.  

[snippet.log.txt](https://github.com/pencilblue/pencilblue/files/659190/snippet.log.txt)

**Comment by efleurine** [comment]
I am not so sure if our problems are related together I just ask a question on stakoverflow which has 2 similarities with yours:
1 - I have this same undefined error and  
2 - I am working on a theme and was taking examples from portfolio  
> the Index.prototype.init of the index controller.

If you look at the code below in // [1] things work correctly but in [2] and beyond the mess started

function Index(){}
	util.inherits(Index, pb.BaseController);


	Index.prototype.init = function (props, cb) {
		const self = this;
    pb.BaseController.prototype.init.call(this, props, function () {
    	const contentService = new pb.ContentService();
    	contentService.getSettings(function(err, contentSettings) {
    		if (util.isError(err)) {
    			cb(err);
    		}

    		// create the service
    		self.contentSettings = contentSettings;
    		const asContext = self.getServiceContext();
    		asContext.contentSettings = contentSettings;
    		self.service = new pb.ArticleServiceV2(asContext);

				// create the loader    	
				const cvlContext								= self.getServiceContext();
				cvlContext.contentSettings 			= contentSettings;	
				cvlContext.service 							= self.service;
				self.contentViewLoader					= new pb.ContentViewLoader(cvlContext);
				 
				// provide a dao
				self.dao = new pb.DAO();

			console.log("\n************"+self.contentSettings.articles_per_page); // [1]
    	});
    	console.log("\n************"+self.contentSettings.articles_per_page); // [2]
    	cb();
    });
 	};

 PB version "0.7.0", and node V5.0.0 on debian jessie

http://stackoverflow.com/questions/41236435/what-i-am-missing-with-these-self-this

**Comment by efleurine** [comment]
I solved my problem and it was about the callback instead of calling in where it is actually I should have called it inside the contentService.getSettings() more specifically where I have the // [1]

**Comment by wu-lee** [comment]
An idea: could my problems be cause by time drift on my server?

I ask since I tested pencilblue on an essentially identical VM (same ISP, same kernel, OS version, similar setup).  I couldn't replicate the problem creating an article.  So what's the difference?

My first hunch was the time, and indeed the time on the misbehaving server is about 4 minutes slow.  It also didn't have ntpd installed (which is probably why it was slow).  The behaving server had ntpd, and wasn't slow.  Finally, my laptop also has ntpd installed and the time is in sync.

Could this be something caused by the differential between the browser's idea of "now" and the server's?

I am currently attempting to replicate the problem on the misbehaving server with the latest master branch of pencilblue, to match the behaving one, then I will install ntpd, sync the time and check again.

**Comment by wu-lee** [comment]
Update:  the odd thing about the misbehaving server is that it is still respawning a job, as reported previously in this issue.  This actually makes the server unusable in itself: connections are reset before any page is rendered. 

The log messages look to me like before: initialising the plugin over and over.  Registering locales, routes for the [pencilblue] theme, creating indexes.   No errors I can see. Repeat about once a second. 

Re-installing from current master (853a46ce68c), pencilblue seems ok until I install the pencilblue plugin.  Then it goes into this state again.

After installing ntpd, and verifying the time was synced, and dropping the database, it seems to be continuing to do this.  Even after a reboot, and dropping the database again.

So maybe it's something else.

**Comment by wu-lee** [comment]
Note. The periodic repeated logs with no connections possible is due to a syntax error in the portfolio plugin's blog.js file, and is my fault.  However, this error is not reported in the logs!  It silently kills off pencilblue's worker thread, which restarts, and so it repeats.

**Comment by wu-lee** [comment]
Tentatively, the site now behaves in that an article created now appears immediately.  Gotta stop now, but I suspect the date lag may have been to blame after all.

**Comment by wu-lee** [comment]
I've attempted to make the site misbehave again by turning off ntpd and setting the time back an hour, with not entirely conclusive results.  I couldn't get it to repeat the behaviour documented above, instead every non-admin page seems to get a 404 not found error (rendered by pencilblue). Restarting, resetting db, etc. doesn't seem to banish it.  I gave up and set the time back, re-installed, and the site is more or less behaving, I do see errors in the console sometimes but the functionality seems to be as I'd expect.

So I think I'd suggest closing this ticket until I can provide something more concrete.

The main suggestion I'd have is that if you think that a drifted server or client times could cause bad behaviour, possibly it'd be worth checking they match up adequately and warning the user if not, as they'd probably be hard to diagnose without a clue like that.

Also - syntax errors in themes should definitely appear as errors in the blog, and they didn't seem to!

Thanks for your help.
